src/resmom/mom_comm.c

```diff
@@ -6156,14 +6524,332 @@ tm_request(int fd, int version)
+					phost->hn_sister = SISTER_EOF;
+					if (phost->hn_stream < 0) {
+						sprintf(log_buffer, "tpp_open failed on %s:%d",
+							phost->hn_host, phost->hn_port);
+						log_joberr(-1, __func__, log_buffer, jobid);
+						continue;
+					}
+				
+					if (pbs_conf.pbs_use_mcast == 1) {
+						/* add each of the tpp streams to the tpp mcast channel */
+						if (tpp_mcast_add_strm(mtfd, phost->hn_stream, TRUE) == -1) {
+							tpp_close(phost->hn_stream);
+							phost->hn_stream = -1;
+							sprintf(log_buffer, "mcast add node %d failed", tvnodeid);
+							log_joberr(-1, __func__, log_buffer, jobid);
+							continue;
+						}
+					}
+					/* Add the IM_SPAWN_MULTI event to the list */	
+					if (nep == NULL) {
+						nep = event_alloc(pjob, IM_SPAWN_MULTI, fd, phost, event, fromtask);
+						ep = nep;
+					} else {
+						ep = event_dup(nep, pjob, phost);
+					}
+					if (ep == NULL) {
+						sprintf(log_buffer, "failed to create event for %s",
+								phost->hn_host?phost->hn_host:"node");
+						log_err(errno, __func__, log_buffer);
+						tpp_close(phost->hn_stream);
+						phost->hn_stream = -1;
+						continue;
+					}	
+					phost->hn_sister = SISTER_OKAY;
+					num++;
+				} /* end of else case */
+			} /* end of looping through the nodes */
+
+			/* Save info off to the job struct */
+			(void)tpp_mcast_members(mtfd, &momcnt);
+			tempinfo->ji_num_sisters = momcnt;
+			tempinfo->ji_taskid_list = temptaskid;
+			tempinfo->ji_nid_list = tempnid;
+			tempinfo->ji_taskid_index = index;
+			
+			if (num > 0) {
+				/* there's info to send to the sisters */
+				ret = im_compose(mtfd, pjob->ji_qs.ji_jobid,
+						get_jattr_str(pjob, JOB_ATR_Cookie),
+						IM_SPAWN_MULTI, ep->ee_event, fromtask,
+						found_empty_string ? IM_PROTOCOL_VER : IM_OLD_PROTOCOL_VER);
+			
+				/* send body of the message: 
+				** start with the identifier,
+				** then the number of items in node_list, 
+				** then send the list of where in a loop
+				** then argc, argv, env, etc.
+				*/
+				if (ret == DIS_SUCCESS)
+					ret = diswui(mtfd, taskident);
+				if (ret == DIS_SUCCESS)
+					ret = diswui(mtfd, list_size);
+				for (i = 0; i < list_size && ret == DIS_SUCCESS; i++) {
+					ret = diswui(mtfd, node_list[i]);
+				}
+				if (ret == DIS_SUCCESS)
+					ret = diswsi(mtfd, myvnodeid);
+				if (found_empty_string && ret == DIS_SUCCESS) {
+					ret = diswui(mtfd, argc);
+					for (i = 0; i < argc && ret == DIS_SUCCESS; i++) {
+						ret = diswst(mtfd, argv[i]);
+					}
+				} else {
+				  	for (i = 0; argv[i] && ret == DIS_SUCCESS; i++) {
+						ret = diswst(mtfd, argv[i]);
+					}
+					if (ret == DIS_SUCCESS)
+						ret = diswst(mtfd, "");
+				}
+				for (i = 0; envp[i] && ret == DIS_SUCCESS; i++) {
+					ret = diswst(mtfd, envp[i]);
+				}
+				if (ret == DIS_SUCCESS) {
+					ret = (dis_flush(mtfd) == -1) ?
+						DIS_NOCOMMIT : DIS_SUCCESS;
+				}
+				if (ret != DIS_SUCCESS) {
+					arrayfree(argv);
```

Comment:
we could consolidate all the frees to the err / done blocks

Reply:
Hmm... I thought about this some more and I think you are right, we should free in the err and done blocks.  I was just following the existing code style, but I think your suggestion is better, so we don't leak memory in the future if code is added that returns without freeing.